chain/events/filter/resolver.go

```diff
@@ -0,0 +1,118 @@
+	AddressResolver func(context.Context, abi.ActorID, *types.TipSet) (address.Address, error)
+	// ActorResolver resolves an address to its actor ID at a given TipSet, or the latest TipSet if nil.
+	ActorResolver func(context.Context, address.Address, *types.TipSet) (abi.ActorID, error)
+
+	tipSetInfo interface {
+		abi.ActorID | address.Address
+	}
+	cachedTipSetInfoResolver[K tipSetInfo, V tipSetInfo] struct {
+		delegate       func(context.Context, K, *types.TipSet) (V, error)
+		cache          *expirable.LRU[string, cachedTipSetInfoResolution[V]]
+		cacheNilTipSet bool
+		keyer          func(K) []byte
+	}
+	cachedTipSetInfoResolution[V tipSetInfo] struct {
+		addr V
+		err  error
+	}
+)
+
+// TODO: separate pool instance with a configurable cap for a more optimal allocation based on keyer.
+// maxPooledBufferCap represents the maximum capacity of the pooled byte buffers.
+// The value 100 is based on the assumption that the longest current key size is f4 address plus 8,
+// the length of TipSet height represented in binary.
+const maxPooledBufferCap = 100
+
+var cacheBufferPool = sync.Pool{
+	New: func() any {
+		return new(bytes.Buffer)
+	},
+}
+
+func NewCachedActorResolver(delegate ActorResolver, size int, expiry time.Duration, cacheNilTipSet bool) ActorResolver {
+	return newCachedTipSetInfoResolver[address.Address, abi.ActorID](
+		delegate,
+		func(addr address.Address) []byte { return addr.Bytes() },
+		size, expiry, cacheNilTipSet)
+}
+
+func NewCachedAddressResolver(delegate AddressResolver, size int, expiry time.Duration, cacheNilTipSet bool) AddressResolver {
+	return newCachedTipSetInfoResolver[abi.ActorID, address.Address](
+		delegate,
+		func(aid abi.ActorID) []byte {
+			buf := make([]byte, 8)
+			binary.BigEndian.PutUint64(buf, uint64(aid))
+			return buf
+		},
+		size, expiry, cacheNilTipSet)
+}
+
+func newCachedTipSetInfoResolver[K tipSetInfo, V tipSetInfo](
+	delegate func(context.Context, K, *types.TipSet) (V, error),
+	keyer func(K) []byte,
+	size int,
+	expiry time.Duration,
+	cacheNilTipSet bool) func(context.Context, K, *types.TipSet) (V, error) {
+	resolver := &cachedTipSetInfoResolver[K, V]{
+		delegate:       delegate,
+		cache:          expirable.NewLRU[string, cachedTipSetInfoResolution[V]](size, nil, expiry),
+		cacheNilTipSet: cacheNilTipSet,
+		keyer:          keyer,
+	}
+	return resolver.resolve
+}
+
+func (c *cachedTipSetInfoResolver[K, V]) resolve(ctx context.Context, k K, ts *types.TipSet) (V, error) {
+	if ts == nil && !c.cacheNilTipSet {
+		// Bypass cache if tipset is set to nil, i.e. the latest.
+		return c.delegate(ctx, k, ts)
+	}
+	key := c.newTipSetInfoKey(k, ts)
+	resolution, found := c.cache.Get(key)
+	if !found {
+		resolution.addr, resolution.err = c.delegate(ctx, k, ts)
+		c.cache.Add(key, resolution)
+	}
+	return resolution.addr, resolution.err
+}
+
+func (c *cachedTipSetInfoResolver[K, V]) newTipSetInfoKey(k K, ts *types.TipSet) string {
+	kk := c.keyer(k)
+	if ts == nil {
+		return string(kk)
```

Comment:
Why do we not use a pooled buffer if tipset is absent ?

Reply:
From design perspective, because `keyer` is beyond the scope of what poole buffer is used for. 

From implementation perspective, using pooling in the keyer won't make much difference since getting a key for an actor ID or address is implemented deep inside those structs. It would entail a larger refactor for little gain.